File: c++/forwardlist.cpp
#include "forwardlist.h"

ForwardListNode::ForwardListNode(std::string key)
{
    this->key = key;
    this->next = nullptr;
}

ForwardList::ForwardList()
{
    this->head = nullptr;
}

// Вставка по индексу
void ForwardList::insert(std::string key, int index)
{
    if(index == 0)
    {
        addHead(key);
        return;
    }
    
    ForwardListNode *node = this->head;
    for(int i = 0; i < index - 1; i++)
    {
        if(node == nullptr)
            break;
        node = node->next;
    }
    if(node == nullptr)
        return;
    ForwardListNode *newNode = new ForwardListNode(key);
    newNode->next = node->next;
    node->next = newNode;

}

// Удаление по индексу
void ForwardList::remove(int index)
{
    if(index == 0)
    {
        removeHead();
        return;
    }

    ForwardListNode *node = this->head;
    for(int i = 0; i < index - 1; i++)
    {
        if(node == nullptr)
            break;
        node = node->next;
    }
    if(node == nullptr)
        return;
    ForwardListNode *toDelete = node->next;
    if(toDelete != nullptr)
    {
        node->next = toDelete->next;
        delete toDelete;
    }
}

// Добавление в хвост
void ForwardList::addTail(std::string key)
{
    if(this->head == nullptr)
    {
        this->head = new ForwardListNode(key);
        return;
    }
    ForwardListNode *node = this->head;
    while(node->next != nullptr)
        node = node->next;
    node->next = new ForwardListNode(key);
}

// Добавление в голову
void ForwardList::addHead(std::string key)
{
    if(this->head == nullptr)
    {
        this->head = new ForwardListNode(key);
        return;
    }
    ForwardListNode *node = new ForwardListNode(key);
    node->next = this->head;
    this->head = node;
}

// Удаление хвоста
void ForwardList::removeTail()
{
    if(this->head == nullptr)
        return;

    ForwardListNode *node = this->head;
    while(node->next->next != nullptr)
        node = node->next;
    delete node->next;
    node->next = nullptr;
}

// Удаление головы
void ForwardList::removeHead()
{
    if(this->head == nullptr)
        return;

    ForwardListNode *toDelete = this->head;
    this->head = toDelete->next;
    delete toDelete;
}

// Вывод с головы
std::string ForwardList::printFromHead()
{
    ForwardListNode *node = this->head;
    if(node == nullptr)
        return "";
    std::string a = "";
    while(node->next != nullptr)
    {
        a += node->key + ' ';
        node = node->next;
    }
    a += node->key;
    return a;
}

// Вывод с хвоста
void ForwardList::printFromTailRec(std::string &s, ForwardListNode *node)
{
    if(node == nullptr)
        return;
    printFromTailRec(s, node->next);
    s += node->key + " ";
}

std::string ForwardList::printFromTail()
{
    std::string s = "";
    printFromTailRec(s, this->head);
    return s;
}

// Удаление элемента по индексу
void ForwardList::removeKey(std::string key, int num)
{
    if(num < 0 || this->head == nullptr)
        return;
    int n = 0;
    ForwardListNode *node = this->head;
    if(node->key == key && num == 0)
    {
        removeHead();
        return;
    }
    while(node->next != nullptr)
    {
        if(node->next->key == key)
        {
            if(n == num)
            {
                ForwardListNode *toDelete = node->next;
                node->next = toDelete->next;
                delete toDelete;
                return;
            }
            n++;
        }
        node = node->next;
    }
}

// Нахождение элемента
ForwardListNode *ForwardList::find(std::string key, int num)
{
    if(num < 0)
        return nullptr;
    int n = 0;
    ForwardListNode *node = this->head;
    while(node != nullptr)
    {
        if(node->key == key)
        {
            if(n == num)
                return node;
            n++;
        }
        node = node->next;
    }
    return nullptr;
}

// Добавление до элемента 
void ForwardList::addBefore(std::string target, std::string key, int occurrence)
{
    if (occurrence < 0 || this->head == nullptr)
        return;

    int n = 0;
    ForwardListNode* prev = nullptr;
    ForwardListNode* current = this->head;

    if (current->key == target && occurrence == 0)
    {
        addHead(key);
        return;
    }

    while (current != nullptr)
    {
        if (current->key == target)
        {
            if (n == occurrence)
            {
                if (prev == nullptr) // голова
                {
                    addHead(key);
                }
                else
                {
                    ForwardListNode* newNode = new ForwardListNode(key);
                    newNode->next = current;
                    prev->next = newNode;
                }
                return;
            }
            n++;
        }
        prev = current;
        current = current->next;
    }
}

// Удаление до элемента
void ForwardList::removeBefore(std::string target, int occurrence)
{
    if (occurrence < 0 || this->head == nullptr || this->head->next == nullptr)
        return;

    int n = 0;
    ForwardListNode* prevPrev = nullptr;
    ForwardListNode* prev = this->head;
    ForwardListNode* current = this->head->next;

    if (prev->key == target && occurrence == 0)
    {
        return;
    }

    while (current != nullptr)
    {
        if (current->key == target)
        {
            if (n == occurrence)
            {
                if (prevPrev == nullptr) // голова
                {
                    removeHead();
                }
                else
                {
                    prevPrev->next = current;
                    delete prev;
                }
                return;
            }
            n++;
        }
        prevPrev = prev;
        prev = current;
        current = current->next;
    }
}

// Добавление после элемента 
void ForwardList::addAfter(std::string target, std::string key, int occurrence)
{
    if (occurrence < 0)
        return;

    int n = 0;
    ForwardListNode* node = this->head;

    while (node != nullptr)
    {
        if (node->key == target)
        {
            if (n == occurrence)
            {
                ForwardListNode* newNode = new ForwardListNode(key);
                newNode->next = node->next;
                node->next = newNode;
                return;
            }
            n++;
        }
        node = node->next;
    }
}

// Удаление после элемента
void ForwardList::removeAfter(std::string target, int occurrence)
{
    if (occurrence < 0 || this->head == nullptr)
        return;

    int n = 0;
    ForwardListNode* node = this->head;

    while (node != nullptr)
    {
        if (node->key == target)
        {
            if (n == occurrence)
            {
                if (node->next != nullptr)
                {
                    ForwardListNode* toDelete = node->next;
                    node->next = toDelete->next;
                    delete toDelete;
                }
                return;
            }
            n++;
        }
        node = node->next;
    }
}

File: go/forwardlist.go
package main

type ForwardListNode struct {
	key  string
	next *ForwardListNode
}

type ForwardList struct {
	head *ForwardListNode
}

func (this *ForwardList) insert(key string, index int) {
	if index == 0 {
		this.addHead(key)
		return
	}

	node := this.head
	for i := 0; i < index-1; i++ {
		if node == nil {
			break
		}
		node = node.next
	}
	if node == nil {
		return
	}
	newNode := &ForwardListNode{key, nil}
	newNode.next = node.next
	node.next = newNode
}

func (this *ForwardList) remove(index int) {
	if index == 0 {
		this.removeHead()
		return
	}

	node := this.head
	for i := 0; i < index-1; i++ {
		if node == nil {
			break
		}
		node = node.next
	}
	if node == nil {
		return
	}
	node.next = node.next.next
}

func (this *ForwardList) addTail(key string) {
	if this.head == nil {
		this.head = &ForwardListNode{key, nil}
		return
	}
	node := this.head
	for node.next != nil {
		node = node.next
	}
	node.next = &ForwardListNode{key, nil}
}

func (this *ForwardList) addHead(key string) {
	if this.head == nil {
		this.head = &ForwardListNode{key, nil}
		return
	}
	node := &ForwardListNode{key, nil}
	node.next = this.head
	this.head = node
}

func (this *ForwardList) removeTail() {
	if this.head == nil {
		return
	}

	node := this.head
	for node.next.next != nil {
		node = node.next
	}
	node.next = nil
}

func (this *ForwardList) removeHead() {
	if this.head == nil {
		return
	}

	this.head = this.head.next
}

func (this *ForwardList) printFromHead() string {
	node := this.head
	if node == nil {
		return ""
	}
	a := ""
	for node.next != nil {
		a += node.key + " "
		node = node.next
	}
	a += node.key
	return a
}

func printFromTailRec(s *string, node *ForwardListNode) {
	if node == nil {
		return
	}
	printFromTailRec(s, node.next)
	*s += node.key + " "
}

func (this *ForwardList) printFromTail() string {
	s := ""
	printFromTailRec(&s, this.head)
	return s
}

func (this *ForwardList) removeKey(key string, num int) {
	if num < 0 || this.head == nil {
		return
	}
	n := 0
	node := this.head
	if node.key == key && num == 0 {
		this.removeHead()
		return
	}
	for node.next != nil {
		if node.next.key == key {
			if n == num {
				node.next = node.next.next
				return
			}
			n++
		}
		node = node.next
	}
}

func (this *ForwardList) find(key string, num int) *ForwardListNode {
	if num < 0 {
		return nil
	}
	n := 0
	node := this.head
	for node != nil {
		if node.key == key {
			if n == num {
				return node
			}
			n++
		}
		node = node.next
	}
	return nil
}

func (this *ForwardList) addBefore(key, targetKey string, num int) {
	if this.head == nil {
		return
	}
	
	if this.head.key == targetKey {
		count := 0
		node := this.head
		for node != nil && node.key == targetKey {
			if count == num {
				newNode := &ForwardListNode{key, nil}
				newNode.next = this.head
				this.head = newNode
				return
			}
			count++
			node = node.next
		}
	}

	count := 0
	prev := this.head
	for prev.next != nil {
		if prev.next.key == targetKey {
			if count == num {
				newNode := &ForwardListNode{key, nil}
				newNode.next = prev.next
				prev.next = newNode
				return
			}
			count++
		}
		prev = prev.next
	}
}

func (this *ForwardList) addAfter(key, targetKey string, num int) {
	if this.head == nil {
		return
	}
	
	count := 0
	node := this.head
	for node != nil {
		if node.key == targetKey {
			if count == num {
				newNode := &ForwardListNode{key, nil}
				newNode.next = node.next
				node.next = newNode
				return
			}
			count++
		}
		node = node.next
	}
}

func (this *ForwardList) removeBefore(targetKey string, num int) {
	if this.head == nil || this.head.next == nil {
		return
	}

	count := 0
	if this.head.next.key == targetKey {
		temp := this.head
		for temp != nil && temp.key == this.head.next.key {
			if count == num {
				this.removeHead()
				return
			}
			count++
			temp = temp.next
		}
	}

	count = 0
	prev := this.head
	for prev.next != nil && prev.next.next != nil {
		if prev.next.next.key == targetKey {
			if count == num {
				prev.next = prev.next.next
				return
			}
			count++
		}
		prev = prev.next
	}
}

func (this *ForwardList) removeAfter(targetKey string, num int) {
	if this.head == nil {
		return
	}
	
	count := 0
	node := this.head
	for node != nil && node.next != nil {
		if node.key == targetKey {
			if count == num {
				node.next = node.next.next
				return
			}
			count++
		}
		node = node.next
	}
}
